sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnOutputReceivers.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.transforms;
+
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.transforms.DoFn.MultiOutputReceiver;
+import org.apache.beam.sdk.transforms.DoFn.OutputReceiver;
+import org.apache.beam.sdk.values.TupleTag;
+import org.joda.time.Instant;
+
+/**
+ * Common {@link OutputReceiver} and {@link MultiOutputReceiver} classes.
+ */
+public class DoFnOutputReceivers {
+  private static class WindowedContextOutputReceiver<T> implements OutputReceiver<T> {
+    DoFn<?, ?>.WindowedContext context;
+    @Nullable TupleTag<T> outputTag;
+    public WindowedContextOutputReceiver(DoFn<?, ?>.WindowedContext context,
```

Comment:
One easy way to broadcast the nature of these adapters is by the good practice of static factory methods. This one, for example would be something like `outputReceiverForLegacyContext` and you might `@Deprecated`.